src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action['type']) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          stateCopy[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
```

Comment:
Instead of deleting each key in a loop, you can simply assign `stateCopy` to an empty object `{}` to clear it. This is more performant and cleaner. You can do this directly inside the `case 'clear':` block.